sql/src/SqlCsv.sk

```diff
@@ -728,75 +688,95 @@ fun replayDiff(
-        tableDiffs,
-        currentDiff,
+      | Some(
+        DiffParsingState(tableDiffs, currentDiff),
       ) if (line.startsWith(":")) ->
+        currentDiff.lines.push(line);
         tableDiffs.push(currentDiff);
         checkpoint = parseCheckpoint(line);
-        break SKStore.CContinue(
-          buildDiffChain(tableDiffs, identity, userFileOpt, checkpoint),
-        )
-
-      | current @ DPBuildingTableDiff(_, currentDiff) if (line == "\t\t") ->
-        currentDiff.!reset = true;
-        current
-
-      | current @ DPBuildingTableDiff(tableDiffs, currentDiff) ->
-        currentDiff.lines.push(line); // capture the raw input for if we end up rejecting the txn
 
-        parseTableChange(userFileOpt, context, currentDiff.table, line) match {
-        | Some(rows) ->
-          currentDiff.rows.extend(rows);
-          current
-        | None() ->
+        rejectedAnyDiff = false;
+        userFileOpt.each(userFile ->
+          SKStore.withContext(privacyContext -> {
+            accessSolver = SKDB.AccessSolver::create(
+              privacyContext,
+              userFile.name,
+            );
+            for (diff in tableDiffs) {
+              if (rejectedAnyDiff) {
+                break void
+              };
+              tableName = diff.table.name.origName;
+              authorCol = SKDB.getAuthorIdx(diff.table);
+              accessCol = SKDB.getAccessIdx(diff.table);
+              for (row in diff.rows) {
+                SKDB.checkUserCanWriteRow(
+                  privacyContext,
+                  accessSolver,
+                  tableName,
+                  authorCol,
+                  accessCol,
+                  row,
+                ) match {
+                | SKDB.AROK() -> void
+                | SKDB.ARError(err) ->
+                  print_error("Error: " + err.toString());
+                  !rejectedAnyDiff = true;
+                  break void
+                }
+              };
+            };
+          })
+        );
+        if (rejectedAnyDiff) {
           writer = mutable Debug.BufferedWriter(print_raw, 4096);
-          tableDiffs.push(currentDiff);
-          for (diff in tableDiffs) {
-            nack(writer, diff)
+          for (tdiff in tableDiffs) {
+            nack(writer, tdiff)
           };
-          mutable DPRejectingTxn(writer)
+          writer.flush();
+          break SKStore.CContinue(None())
+        } else {
+          break SKStore.CContinue(
+            buildDiffChain(tableDiffs, identity, userFileOpt, checkpoint),
+          )
         }
 
-      | current @ DPRejectingTxn(writer) if (line.startsWith("^")) ->
-        (table, snapshot) = parseHeader(context, line);
-        nack(writer, TableDiff::create(table, snapshot));
+      | current @ Some(DiffParsingState(_, currentDiff)) if (line == "\t\t") ->
+        currentDiff.!reset = true;
         current
 
-      | DPRejectingTxn(writer) if (line.startsWith(":")) ->
-        writer.write(line);
```

Comment:
I think we've lost this? The checkpoint is required for any nack'd lines to be applied and stored. So I think - might be missing it - that we're rejecting and sending lines back to the client, but they'll never be stored in the rejected changes table because the checkpoint is missing to apply them. Can you test this scenario?

Reply:
It's just moved! To line 722.

Reply:
I have tested this scenario -- rejection/feedback works as before, and it's covered by `test_replication_privacy_rejection_and_feedback`

Reply:
722 is `currentDiff.lines.push(line);`? Where is it actually written to stdout?

Reply:
It's nack'ed at line 762 if the transaction is rejected, and written to stdout during the call to `buildDiffChain` otherwise

Reply:
The nack just takes care of the table. We need to call that for each table and then finally write the checkpoint line. It's this last bit that's missing. It will catch up and do the right thing when something eventually succeeds which might be why a test is passing?

Reply:
(closing the loop here for future reference: the current setup is correct/fine; the `nack` takes care of the checkpoint line)